### PR TITLE
[FLINK-4288] [table] Make it possible to unregister tables

### PR DESCRIPTION
**Pull request by twalthr**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.
- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

This PR adds the possibility to unregister tables. Although Calcite supports mutable schemas, deleting a table instead of replacing it was not easy to implement. I will add some documentation if you are fine with this PR.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/FlinkSchema.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import org.apache.calcite.jdbc.CalciteSchema
+import org.apache.calcite.jdbc.CalciteSchema.TableEntry
+import org.apache.calcite.schema
+import org.apache.calcite.schema.SchemaPlus
+
+/**
+  * Wraps [[CalciteSchema]] and allows for deleting tables.
+  */
+class FlinkSchema {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/FlinkSchema.scala]: I think the benefit of this wrapper is rather limited.
Couldn't we just call `schema.tableMap.remove()` in `TableEnvironment`?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala [hunk 2]
```diff
@@ -133,12 +134,24 @@ abstract class TableEnvironment(val config: TableConfig) {
...
 
   /**
+    * Unregisters a [[Table]] in the TableEnvironment's catalog.
+    * Unregistered tables cannot be referenced in SQL queries anymore.
+    *
+    * @param name The name under which the table is registered.
+    */
+  def unregisterTable(name: String): Unit = {
+
+    checkValidTableName(name)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala]: Do we need to check if the name is valid? Couldn't we just try to delete it?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala [hunk 3]
```diff
@@ -133,12 +134,24 @@ abstract class TableEnvironment(val config: TableConfig) {
...
 
   }
 
   /**
+    * Unregisters a [[Table]] in the TableEnvironment's catalog.
+    * Unregistered tables cannot be referenced in SQL queries anymore.
+    *
+    * @param name The name under which the table is registered.
+    */
+  def unregisterTable(name: String): Unit = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala]: Do we want to return a boolean success value or throw an exception if the table does not exist?

**Comment by twalthr** [comment]
I updated this PR. I will create a followup issue to refactor all TableEnvironment registration tests to unit tests.

**Comment by fhueske** [comment]
Thanks for the update. +1 to merge.

**Comment by twalthr** [comment]
Thanks @fhueske. I will add some sentences to the documentation and merge this.

**Pull request closed without merging**